util/webhook/certs/certmanager.go

```diff
@@ -0,0 +1,275 @@
+func (cm *CertificateManager) start(ctx context.Context) error {
+	var rotationTime time.Time
+	for {
+		nextRotationTime, err := cm.rotateCerts(ctx, rotationTime)
+		if err != nil {
+			return err
+		}
+		rotationTime = nextRotationTime
+		select {
+		case <-time.After(time.Until(nextRotationTime)):
+		case <-ctx.Done():
+			return nil
+		}
+	}
+}
+
+// rotateCerts rotates certificates at the scheduled rotation time, writes them to the secret and returns the next
+// scheduled rotation time.
+func (cm *CertificateManager) rotateCerts(ctx context.Context, scheduledNextRotationTime time.Time) (time.Time, error) {
+	now := time.Now()
+	cm.opts.Logger.Info("Rotating certificates", "now", now.String(), "scheduledNextRotationTime", scheduledNextRotationTime.String())
+
+	// Get the webhook server's secret. Certificate manager expects an empty secret to have been created with the
+	// controller deployment that intends to use cert management.
+	// Any error in getting that secret is cause for exiting this program.
+	secret := &corev1.Secret{}
+	secretKey := client.ObjectKey{Name: cm.opts.SecretName, Namespace: cm.opts.SecretNamespace}
+	cm.opts.Logger.Info("Getting secret", "namespacedName", secretKey.String())
+	if err := cm.opts.Client.Get(ctx, secretKey, secret); err != nil {
+		return time.Time{}, err
+	}
+
+	if len(secret.Annotations) == 0 {
+		secret.Annotations = map[string]string{}
+	}
+
+	// If this is the first time rotate is called since starting certificate manager, then initialize the webhooks with
+	// any CA data (if any) in the secret. This is for the case when a package is redeployed (for any reason) and the
+	// webhook config properties are reset.
+	if scheduledNextRotationTime.IsZero() {
+		if caCertData := secret.Data[CACertName]; len(caCertData) > 0 {
+			if err := cm.updateWebhookConfigs(ctx, caCertData); err != nil {
+				return time.Time{}, err
+			}
+		}
+	}
+
+	// If the webhook secret already has a scheduled next rotation time that does not occur in the past, then defer this
+	// rotation until that time. This usually happens when another certificate manager has updated the certs (perhaps
+	// from an earlier process that was exited).
+	nextRotationTime, _ := cm.getNextRotationTime(secret)
+	if !nextRotationTime.IsZero() && !nextRotationTime.Before(now) {
+		if !scheduledNextRotationTime.Equal(nextRotationTime) {
+			cm.opts.Logger.Info("Rescheduling next rotation", "nextRotationTime", nextRotationTime.String())
+			return nextRotationTime, nil
+		}
+	}
+
+	// Determine the rotation interval.
+	var rotationInterval time.Duration
+	if value := secret.Annotations[cm.opts.RotationIntervalAnnotationKey]; value != "" {
+		if rotationInterval, _ = time.ParseDuration(value); rotationInterval == 0 {
+			rotationInterval = defaultRotationInterval
+		}
+	}
+
+	// Update next rotation time and write it to the NextRotationAnnotationKey in the secret.
+	nextRotationTime = now.Add(rotationInterval)
```

Comment:
If annotation value is empty, should we set nextRotationTime as now + defaultRotationInterval ? Looking at L109 only executes if annotation has some value, otherwise L108 would render rotationInterval as 0s.

Reply:
Good catch. I fixed this by removing the `value != ""` check.

Reply:
Avi's argument holds true if the annotation itself is not present?

Reply:
That is also taken care of by the way Go handles missing keys in maps.

```go
value := secret.Annotations[cm.opts.RotationIntervalAnnotationKey]
```

`value` will have a zero-value (`""`) if the key is missing and `rotationInterval` will be parsed to `0`.